include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1012 @@
+    bool
+    empty() const noexcept
+    {
+        return n_ == 0;
+    }
+
+    /** Copies characters
+
+        Copies a substring to the character array pointed to by
+        the destination char string, where `rcount` is the smaller
+        of `count` and `size() - pos`.
+
+        @par Exception Safety
+        Strong guarantee.
+        Exceptions thrown on invalid positions.
+
+        @throw out_of_range If `pos > size()`
+
+        @param dest pointer to the destination character string
+        @param count requested substring length
+        @param pos position of the first character
+
+        @return Number of characters copied
+    */
+    BOOST_URL_DECL
+    size_type
+    copy(char* dest,
+         size_type count,
+         size_type pos = 0) const;
+
+    /** Compares two strings
+
+        The length of the sequences to compare is the smaller of
+        `size()` and `other.size()`.
+
+        The function compares the two strings as if by calling
+        `char_traits<char>::compare(to_string().data(), v.data(), rlen)`.
+
+        @param other string to compare
+
+        @return Negative value if this string is less than the other
+        character sequence, zero if the both character sequences are
+        equal, positive value if this string is greater than the other
+        character sequence
+    */
+    BOOST_URL_DECL
+    int
+    compare(string_view other) const;
+
+    /// @copydoc compare(string_view)
+    int
+    compare(char const* other) const
+    {
+        return compare(string_view(other));
+    }
+
+    /** Compares two strings
+
+        The length of the sequences to compare is the smaller of
+        `size()` and `other.size()`.
+
+        The function compares the two strings as if by calling
+        `char_traits<char>::compare(to_string().data(), v.to_string().data(), rlen)`.
+
+        @param other string to compare
+
+        @return Negative value if this string is less than the other
+        character sequence, zero if the both character sequences are
+        equal, positive value if this string is greater than the other
+        character sequence
+    */
+    BOOST_URL_DECL
+    int
+    compare(decoded_view other) const;
+
+    /** Returns a basic_string
+
+        This overload allows the caller to recycle
+        the input string.
+
+        @return A string representing the
+        entire contents of the decoded range.
+    */
+    template <class A>
+    std::basic_string<char, std::char_traits<char>, A>&
+    to_string(std::basic_string<char, std::char_traits<char>, A> & str) const
+    {
+        str.resize(size());
+        copy(&str[0], size());
+        return str;
+    }
+
+    /** Returns a basic_string
+
+        This overload constructs a new string.
+
+        @return A string representing the
+        entire contents of the decoded range.
+    */
+    template <class A = std::allocator<char>>
+    std::basic_string<char, std::char_traits<char>, A>
+    to_string() const
```

Comment:
No no no... too much try-hard here! Users don't want to have to write `x.template to_string<Y>()` this should just return `std::string`

Reply:
They don't have to write `x.template to_string<Y>()`.

https://github.com/alandefreitas/url/blob/9c576efadbc653faef31a7f827a5182d462fd488/test/unit/decoded_view.cpp#L331

Reply:
Sometimes you do but in any event the signature should be just `std::string decoded_view::to_string() const` and it can be inline (calling assign_to on a returned local)

Reply:
https://godbolt.org/z/7sT4T487P

The general rule is that we try to avoid creating APIs where the user passes a template parameter to a member function (think `std::get`). Sometimes it is OK but it is rare, this is not one of those rare cases.